azure_gateway/_gateway_llm.py

```diff
@@ -0,0 +1,256 @@
+import json
+import logging
+import os
+import time
+from typing import Callable
+
+import requests
+import tiktoken
+from langchain_core.language_models.llms import LLM
+from openai import AzureOpenAI
+from requests_sse import EventSource
+
+from azure_gateway import parse_openai_chat_response, OutOfQuotaError
+from azure_gateway._logger import LoggerWrapper
+
+
+Callback = Callable[[str], None]
+ServiceResponse = tuple[str, int, int]
+
+
+class BaseGatewayLLM(LLM):
+    """Base class for Gateway LLM"""
+    project_id: str
+    model_id: str
+    model_type: str
+    conversation_history: list
+    gateway_auth_token: str
+    alert_log_threshold: float = 0.9
+    logger: logging.Logger | None = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.log = LoggerWrapper(self.logger)
+
+    @property
+    def _llm_type(self) -> str:
+        return "custom"
+
+    @property
+    def _identifying_params(self):
+        """Get the identifying parameters."""
+        return {}
+
+    def _call(self, prompt: str, callback: Callback | None, **_):
+        self._query_gateway_safely(query_function=self._check_quota)
+
+        self.log.info(f"PROMPT\n{self.log.indent_text(prompt)}")
+        self.conversation_history.append({"role": "user", "content": prompt})
+
+        response_text, prompt_tokens, completion_tokens = self._call_service(callback=callback)
+
+        self._query_gateway_safely(
+            query_function=lambda: self._update_quota(
+                prompt_tokens=prompt_tokens, completion_tokens=completion_tokens
+            )
+        )
+        self.log.info(f"RESPONSE\n{self.log.indent_text(response_text)}")
+
+        self.conversation_history.append({"role": "assistant", "content": response_text})
+
+        return response_text
+
+    def _call_service(self, callback: Callback | None) -> ServiceResponse:
+        raise NotImplementedError()
+
+    def call(self, prompt: str, detail: str, callback: Callback | None) -> str:
+        """Wrap-around for Langchain's invoke method."""
+        start = time.time()
+        response = self.invoke(input=prompt, config=None, callback=callback)
+        self.log.info(f"CALLTIME - {detail} - {time.time() - start}")
+        return response
+
+    def _check_quota(self):
+        """Check the current usage quota and raise an error if the project is out of quota."""
+        quota_check_response = requests.get(
+            os.environ["GATEWAY_HIGH_TRAFFIC_URL"],
```

Comment:
Should this be passed as a variable? It makes the projects dependent on the value without being obvious.